lib/fixed-points.nix

```
@@ -311,63 +316,141 @@ rec {
+    in
+    fApplied // g final prev';
+
+  /**
+    Composes a list of [`overlays`](#chap-overlays) and returns a single overlay function that combines them.
 
+    :::{.note}
+    The result is produced by using the update operator `//`.
+    This means nested values of previous overlays are not merged recursively.
+    :::
 
     # Inputs
 
-    `f`
+    `extensions`
 
-    : 1\. Function argument
+    : A list of overlay functions
+      :::{.note}
+      The order of the overlays in the list is important.
+      :::
 
-    `g`
-
-    : 2\. Function argument
+    : Each overlay function takes two arguments, `final` and `prev`, and returns an attribute set.
+      - `final` is the result of the fixed-point function, with all overlays applied.
+      - `prev` is the result of the previous overlay function.
 
     `final`
 
-    : 3\. Function argument
+    : The return value of the original function that is extended by the overlays `f` and `g`.
+
+      If `composeManyExtensions extensions` is itself used in a composition of overlays, then any overlays that precede it can be considered to be part of "the original function".
 
     `prev`
 
-    : 4\. Function argument
-  */
-  composeExtensions =
-    f: g: final: prev:
-      let fApplied = f final prev;
-          prev' = prev // fApplied;
-      in fApplied // g final prev';
+    : The intermediate result of applying preceding overlays.
 
-  /**
-    Compose several extending functions of the type expected by 'extends' into
-    one where changes made in preceding functions are made available to
-    subsequent ones.
+    # Type
 
     ```
-    composeManyExtensions : [packageSet -> packageSet -> packageSet] -> packageSet -> packageSet -> packageSet
-                              ^final        ^prev         ^overrides     ^final        ^prev         ^overrides
+    # Pseudo code
+    let
+      OverlayFn :: ( { ... } -> { ... } -> { ... } );
+    in
+      composeManyExtensions :: [ OverlayFn ] -> OverlayFn
     ```
+
+    # Examples
+    :::{.example}
+    ## `lib.fixedPoints.composeManyExtensions` usage example without `lib.extends`
+
+    The following usage of `lib.fixedPoints.composeManyExtensions` without `lib.extends` demonstrates overlays as an atomic concept.
+    In nixpkgs however, it is required to use `lib.extends` such that the original function becomes part of the fixed point as shown in the next example.
+
+    ```nix
+    let
+      # Each overlay function has 'final' and 'prev' as arguments.
+      overlayA = final: prev: {
+        b = final.c; # The final value of c after overlayB
+        c = 3;
+      };
+      overlayB = final: prev: {
+        c = 10;
+        cPrev = prev.c; # The value of c before overlayB
+      };
+
+      extensions = composeManyExtensions [ overlayA overlayB ];
+
+      # Calculate the fixed point of all composed overlays.
+      # With an empty attribute set as the initial value.
+      fixedpoint = lib.fix (final: extensions final {});
```

I don't think we should do this, because it is easy to change it and get it wrong.
When a use case has a clear separation (ie of responsibilities) between a base layer and e.g. user overlays, it is tempting to put the base layer in place of `{ }`, but this is wrong because it won't be included in the fixed point, as in `baseLayer // overlayStuff`.
If we use `extends` here, this mistake won't be made.
```suggestion
      # Calculate the fixed point of all composed overlays.
      # With an empty attribute set as the initial value.
      fixedpoint = lib.fix (lib.extends extensions {});
```

---

Oh, I guess the next section has it. Maybe just remove this variation then?

---

I would like to keep this example. The benefit of having this example is, that is doesn't require understanding extends, which causes additional brainload and can be complicated. I thought of this example as something that works at the bare minimum required for understanding the concept.

> [!WARNING]  
> This function doesn't include the initial attribute set into the fixed point.
> For a real world scenario the next example shows the correct usage.

---

Ok. Agreed. api/reference documentation should contain one minimal working example.